File: sdk/dotnet/src/Infergo/InfergoException.cs
namespace Infergo;

/// <summary>
/// Exception thrown when an infergo native call fails.
/// </summary>
public class InfergoException : Exception
{
    /// <summary>
    /// The native error code returned by the C API, or -1 if unavailable.
    /// </summary>
    public int ErrorCode { get; }

    public InfergoException(string message) : base(message)
    {
        ErrorCode = -1;
    }

    public InfergoException(string message, int errorCode)
        : base(message)
    {
        ErrorCode = errorCode;
    }

    public InfergoException(string message, Exception inner)
        : base(message, inner)
    {
        ErrorCode = -1;
    }

    /// <summary>
    /// Build an exception from the native thread-local error string.
    /// </summary>
    internal static InfergoException FromNative(string context, int errorCode = -1)
    {
        string? nativeMsg = Native.GetLastError();
        string msg = string.IsNullOrEmpty(nativeMsg)
            ? context
            : $"{context}: {nativeMsg}";
        return new InfergoException(msg, errorCode);
    }
}

File: sdk/java/src/main/java/com/infergo/InfergoException.java
package com.infergo;

/**
 * Exception thrown by infergo native operations.
 * Wraps error codes and messages from the C API.
 */
public class InfergoException extends RuntimeException {

    private final int errorCode;

    public InfergoException(String message) {
        super(message);
        this.errorCode = -1;
    }

    public InfergoException(int errorCode, String message) {
        super(message);
        this.errorCode = errorCode;
    }

    public InfergoException(String message, Throwable cause) {
        super(message, cause);
        this.errorCode = -1;
    }

    /**
     * Returns the native error code, or -1 if not available.
     */
    public int getErrorCode() {
        return errorCode;
    }
}
